### PR TITLE
Expand space-filling designs

### LINKED ISSUE #307: Revise space-filling designs
We currently make our space-filling designs on-the-fly using functions in `DiceDesign`. They use random numbers and don't always produce the best results. 

[This website](https://www.spacefillingdesigns.nl) has compiled designs that have been optimized for varying numbers of parameters and sizes. They are free of charge and I've already scraped them.

I'd like to: 

- For `grid_latin_hypercube()`, use the [Audze-Eglais designs](https://www.sciencedirect.com/science/article/abs/pii/S0965997803000425) whenever possible (if they are available for the chosen number of parameters and size). Otherwise, do what we've done before. 
- Add a new `grid_max_min()` function to look up the L2 and L1 designs. We might consider moving the default grid in tune to use this. It goes up to 10 parameters so we are probably fine. 

This is also made me feel that we should adjust the `grid_max_entropy()` function to raise the values of the radius and the number of iterations to make them more stable. We could also add a `tries` argument that will use different random numbers and take the design with the best results.

### PR DESCRIPTION
**Pull request by topepo**
Closes #307

The space-filling design functionality was expanded to include several new types of designs: Audze-Eglais, max/min L1, max/min L2, and uniform. These are all pre-computed designs accessed from the sfd package. 

A new function is used to access all of the space-filling designs called `grid_space_filling()`.

### CODE REVIEW

**Review by hfrick — COMMENTED** [review]
I like one function for space-filling designs better than 6 different ones 👍

With that we have a mismatch in what's available/exported between different types of space-filling designs though: we have `grid_latin_hypercube()` and `grid_max_entropy()` but no, say, `grid_audze_eglais()`. 

I'm in favor of deprecating `grid_latin_hypercube()` and `grid_max_entropy()`, starting with a soft-deprecation. This can and should be a separate issue and PR but we should agree on what to do with that mismatch before adding `grid_space_filling()`.

File: R/space_filling.R [hunk 1]
```diff
@@ -41,6 +49,191 @@
...
+    has_premade_design <- sfd::sfd_available(p, size, type)
+
+    if (has_premade_design) {
+      grid <- sfd::get_design(p, num_points = size, type = type)
+      # TODO value_seq should keep factor levels
+      vals <- purrr::map(params, ~ value_seq(.x, size))
+      vals <- purrr::map(vals, ~ base_recycle(.x, size))
+      grid <- sfd::update_values(grid, vals)
+      names(grid) <- names(params)
+    } else {
```
Comment (topepo) [inline R/space_filling.R:198]: @hfrick Here's where it falls back to a ME design if there is no pre-made design. I can update the wording in the man file to be more specific.

File: R/space_filling.R [hunk 2]
```diff
@@ -2,23 +2,24 @@
...
 #'
 #' @inheritParams grid_random
-#' @param size A single integer for the total number of parameter value
+#' @param size A single integer for the maximum number of parameter value
 #' combinations returned. If duplicate combinations are
 #' generated from this size, the smaller, unique set is returned.
+#' @param type A character string with possible values: `"any"`,
+#' `"audze_eglais"`, `"max_min_l1"`, `"max_min_l2"`, `"uniform"`,
+#' `"max_entropy"`, or `"latin_hypercube"`. A value of `"any"` will choose the
+#' first design available (in the order listed above).
```
[the remark this answers is not in the document]
Comment (topepo) [reply R/space_filling.R:17]: > A value of `"any"` will choose the first design available (in the order listed above).

That's true; there is a branch that produces a ME design if one is not available in sfd (see comment below)

File: R/space_filling.R [hunk 3]
```diff
@@ -41,6 +49,191 @@
...
+  param_quos <- quos(...)
+  params <- map(param_quos, eval_tidy)
+  p <- length(params)
+
+  if (type %in% c("any", "audze_eglais", "max_min_l1", "max_min_l2", "uniform")) {
+    has_premade_design <- sfd::sfd_available(p, size, type)
+
+    if (has_premade_design) {
+      grid <- sfd::get_design(p, num_points = size, type = type)
+      # TODO value_seq should keep factor levels
```
[the remark this answers is not in the document]
Comment (topepo) [reply R/space_filling.R:193]: Now that I look, `?value_validate()` says: 

> `value_seq()` and `value_sample()` return a vector of values consistent with the type field of object.

and it does: 

``` r
library(dials)
#> Loading required package: scales
suppressPackageStartupMessages(library(dplyr))

activation()
#> Activation Function  (qualitative)
#> 5 possible values include:
#> 'linear', 'softmax', 'relu', 'elu' and 'tanh'
activation()$type
#> [1] "character"

value_sample(activation(), 3) %>% class()
#> [1] "character"
value_sample(activation(), 3) %>% class()
#> [1] "character"

grid_regular(activation(), levels = 3)
#> # A tibble: 3 × 1
#>   activation
#>   <chr>     
#> 1 linear    
#> 2 softmax   
#> 3 relu
```

<sup>Created on 2024-04-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Still want an issue?

**Comment by hfrick** [comment]
The failure on R 3.6 is okay, we don't support that version anymore.

**Pull request merged**